### DISCUSSION TITLE
WASM module load failed: unsupported opcode fc 0a

### DISCUSSION THREAD

**Discussion by Zghw2012** [bytecodealliance/wasm-micro-runtime#1797] [Q&A] [answered]
Hello,

I start to use wasm-micro-runtime.  However I can not run the sample.
After I install the wasi-sdk and clone wasm-micro-runtime into local machine (Ubuntu 18). I meet the following problem when I execute the ./multi_module under /wasm-micro-runtime/samples/multi-module/build.

- wasm_runtime_full_init
- wasm_runtime_set_module_reader
- bh_read_file_to_buffer ./mC.wasm
- wasm_runtime_load
- bh_read_file_to_buffer ./mB.wasm
- bh_read_file_to_buffer ./mA.wasm
**WASM module load failed: unsupported opcode fc 0a**
- release the read file buffer
- wasm_runtime_destroy
- release the read file buffer
- release the read file buffer

Any suggestions to solve it?

**Comment by Zghw2012** [comment]
wasm-micro-runtime/samples/multi-module/build> make                                                                                      
 [  2%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/shared/platform/linux/platform_init.c.o
[  4%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/shared/platform/common/posix/posix_malloc.c.o
[  6%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/shared/platform/common/posix/posix_memmap.c.o
[  8%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/shared/platform/common/posix/posix_socket.c.o
[ 10%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/shared/platform/common/posix/posix_thread.c.o
[ 12%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/shared/platform/common/posix/posix_time.c.o
[ 14%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/shared/mem-alloc/ems/ems_alloc.c.o
[ 17%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/shared/mem-alloc/ems/ems_hmu.c.o
[ 19%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/shared/mem-alloc/ems/ems_kfc.c.o
[ 21%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/shared/mem-alloc/mem_alloc.c.o
[ 23%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/shared/utils/bh_assert.c.o
[ 25%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/shared/utils/bh_common.c.o
[ 27%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/shared/utils/bh_hashmap.c.o
[ 29%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/shared/utils/bh_list.c.o
[ 31%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/shared/utils/bh_log.c.o
[ 34%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/shared/utils/bh_queue.c.o
[ 36%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/shared/utils/bh_vector.c.o
[ 38%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/shared/utils/runtime_timer.c.o
[ 40%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/iwasm/libraries/libc-builtin/libc_builtin_wrapper.c.o
[ 42%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c.o
[ 44%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c.o
[ 46%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/random.c.o
[ 48%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/str.c.o
[ 51%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/iwasm/common/wasm_application.c.o
[ 53%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/iwasm/common/wasm_c_api.c.o
[ 55%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/iwasm/common/wasm_exec_env.c.o
[ 57%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/iwasm/common/wasm_memory.c.o
[ 59%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/iwasm/common/wasm_native.c.o
[ 61%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/iwasm/common/wasm_runtime_common.c.o
[ 63%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/iwasm/common/wasm_shared_memory.c.o
[ 65%] Building ASM object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/iwasm/common/arch/invokeNative_em64.s.o
[ 68%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/iwasm/interpreter/wasm_interp_classic.c.o
[ 70%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/iwasm/interpreter/wasm_loader.c.o
[ 72%] Building C object CMakeFiles/vmlib.dir/repo/wasm-micro-runtime/core/iwasm/interpreter/wasm_runtime.c.o
[ 74%] Linking C static library libvmlib.a
[ 74%] Built target vmlib
[ 76%] Creating directories for 'WASM_MODULE'
[ 78%] No download step for 'WASM_MODULE'
[ 80%] No update step for 'WASM_MODULE'
[ 82%] No patch step for 'WASM_MODULE'
[ 85%] Performing configure step for 'WASM_MODULE'
-- The C compiler identification is Clang 14.0.4
-- The CXX compiler identification is Clang 14.0.4
System is unknown to cmake, create:
Platform/WASI to use this system, please post your config file on discourse.cmake.org so it can be added to cmake
-- Detecting C compiler ABI info
System is unknown to cmake, create:
Platform/WASI to use this system, please post your config file on discourse.cmake.org so it can be added to cmake
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /repo/wasi-sdk/build/wasi-sdk-16.5ga0a342ac182c//bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
System is unknown to cmake, create:
Platform/WASI to use this system, please post your config file on discourse.cmake.org so it can be added to cmake
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /repo/wasi-sdk/build/wasi-sdk-16.5ga0a342ac182c//bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting WABT
-- Detecting WABT - found
-- Detecting WASM_OBJDUMP at WABT_DIR-NOTFOUND/wabt
-- Detecting WASM_OBJDUMP at WABT_DIR-NOTFOUND/wabt - not found
-- Detecting WASM2WAT at WABT_DIR-NOTFOUND/wabt
-- Detecting WASM2WAT at WABT_DIR-NOTFOUND/wabt - not found
-- Generating mA.wasm as REACTOR...
-- Generating mB.wasm as REACTOR...
-- Generating mC.wasm as COMMAND...
-- Generating mD.wasm as COMMAND...
-- Generating mE.wasm as REACTOR...
-- Configuring done
-- Generating done
-- Build files have been written to: /repo/wasm-micro-runtime/samples/multi-module/build/WASM_MODULE-prefix/src/WASM_MODULE-build
[ 87%] Performing build step for 'WASM_MODULE'
[ 10%] Building C object CMakeFiles/MODULE_mA.dir/mA.c.obj
[ 20%] Linking C executable mA.wasm
[ 20%] Built target MODULE_mA
[ 30%] Building C object CMakeFiles/MODULE_mB.dir/mB.c.obj
[ 40%] Linking C executable mB.wasm
[ 40%] Built target MODULE_mB
[ 50%] Building C object CMakeFiles/MODULE_mC.dir/mC.c.obj
[ 60%] Linking C executable mC.wasm
[ 60%] Built target MODULE_mC
[ 70%] Building CXX object CMakeFiles/MODULE_mD.dir/mD.cpp.obj
[ 80%] Linking CXX executable mD.wasm
[ 80%] Built target MODULE_mD
[ 90%] Building CXX object CMakeFiles/MODULE_mE.dir/mE.cpp.obj
[100%] Linking CXX executable mE.wasm
[100%] Built target MODULE_mE
[ 89%] Performing install step for 'WASM_MODULE'
[ 91%] Completed 'WASM_MODULE'
[ 91%] Built target WASM_MODULE
[ 93%] Building C object CMakeFiles/multi_module.dir/src/main.c.o
[ 95%] Building C object CMakeFiles/multi_module.dir/repo/wasm-micro-runtime/core/shared/utils/uncommon/bh_getopt.c.o
[ 97%] Building C object CMakeFiles/multi_module.dir/repo/wasm-micro-runtime/core/shared/utils/uncommon/bh_read_file.c.o
[100%] Linking C executable multi_module
[100%] Built target multi_module
/repo/wasm-micro-runtime/samples/multi-module/build> ls
CMakeCache.txt  CMakeFiles  Makefile  WASM_MODULE-prefix  cmake_install.cmake  libvmlib.a  mA.wasm  mB.wasm  mC.wasm  mD.wasm  mE.wasm  multi_module

**Comment by loganek** [comment, marked as answer]
Hi @Zghw2012 , `fc 0a` opcode is for `memory.copy` operation (see https://pengowray.github.io/wasm-ops/) which is part of bulk memory proposal. This is supported by WAMR but only enabled by default very recently: https://github.com/bytecodealliance/wasm-micro-runtime/pull/1781
You either migrate to the latest WAMR, or explicitly enable bulk memory by adding `-DWAMR_BUILD_BULK_MEMORY=1` to your cmake command.

**Reply by Zghw2012** [reply]
Thank you very much!

**Reply by Falzz1010** [reply]
Woww
